src/testing/run.ts

```diff
@@ -0,0 +1,282 @@
+      } catch (err) {
+        // Ignore, errors are handled in the tasks themselves
+      }
+      return { id };
+    })();
+
+    promises.push({
+      id,
+      promise,
+    });
+
+    if (promises.length >= args.maxConcurrency) {
+      // Remove the first promise that resolves
+      await Promise.race(promises.map((p) => p.promise)).then((winner) => {
+        const toRemove = promises.find((p) => p.id === winner.id);
+        if (toRemove) {
+          promises.splice(promises.indexOf(toRemove), 1);
+        }
+      });
+    }
+  }
+
+  await Promise.allSettled(promises.map((p) => p.promise)); // Ensure all remaining promises are finished
+}
+
+async function sendError(args: {
+  testId: string;
+  testCaseHash: string | null;
+  evaluatorId: string | null;
+  error: unknown;
+}): Promise<void> {
+  if (args.error instanceof Error) {
+    await client.post('/errors', {
+      testExternalId: args.testId,
+      testCaseHash: args.testCaseHash,
+      evaluatorExternalId: args.evaluatorId,
+      error: {
+        name: args.error.name,
+        message: args.error.message,
+        stacktrace: args.error.stack,
+      },
+    });
+  }
+}
+
+async function evaluateOutput<TestCaseType, OutputType>(args: {
+  testId: string;
+  testCase: TestCaseType;
+  testCaseHash: string;
+  output: OutputType;
+  evaluator: BaseTestEvaluator<TestCaseType, OutputType>;
+}): Promise<void> {
+  let evaluation: Evaluation | undefined = undefined;
+
+  try {
+    evaluation = await args.evaluator.evaluateTestCase({
+      testCase: args.testCase,
+      output: args.output,
+    });
+  } catch (err) {
+    await sendError({
+      testId: args.testId,
+      testCaseHash: args.testCaseHash,
+      evaluatorId: args.evaluator.id,
+      error: err,
+    });
+  }
+
+  if (evaluation === undefined) return;
+
+  await client.post('/evals', {
+    testExternalId: args.testId,
+    testCaseHash: args.testCaseHash,
+    evaluatorExternalId: args.evaluator.id,
+    score: evaluation.score,
+    threshold: evaluation.threshold,
+  });
+}
+
+async function runTestCase<TestCaseType, OutputType>(args: {
+  testId: string;
+  testCase: TestCaseType;
+  testCaseHash: string;
+  evaluators: BaseTestEvaluator<TestCaseType, OutputType>[];
+  fn: (testCase: TestCaseType) => OutputType | Promise<OutputType>;
+  maxEvaluatorConcurrency: number;
+}): Promise<void> {
+  let output: OutputType | undefined = undefined;
+
+  try {
+    output = await args.fn(args.testCase);
+  } catch (err) {
+    await sendError({
+      testId: args.testId,
+      testCaseHash: args.testCaseHash,
+      evaluatorId: null,
+      error: err,
+    });
+  }
+
+  if (output === undefined) return;
+
+  await client.post('/results', {
+    testExternalId: args.testId,
+    testCaseHash: args.testCaseHash,
+    testCaseBody: args.testCase,
+    testCaseOutput: isPrimitive(output) ? output : JSON.stringify(output),
+  });
+
+  try {
+    await gatherWithMaxConcurrency({
```

Comment:
So max eval concurrency is really max eval concurrency per test case not in total?

Reply:
yeah

Reply:
we may want to revisit that in the future, my initial thinking was that it was in total because the reason I would use it is to avoid rate limits from my llm provider so its a bit hard to do that if it's per test case. we could also just call that out that max eval concurrency is per test case so if you have a maxTestCase concurrency of 5 and a maxEvalConcurrency of 5 you could be running 25 evals in parallel.

Reply:
yeah i have in the docs that it's "per test case"

> How many evaluators to run concurrently **on the result of a test case**.

but open to rewording

Reply:
i also don't think it's hard to do if it's per test case right? can't you choose values that always keeps you under a fixed amount of concurrent reqs?

Reply:
well it scales with test case concurrency. So if I have `max_test_case_concurrency=5` my options for concurrency for evals is either 5,10,15,20 etc

Reply:
yeah that's true, i'll look at making it a global max

Reply:
I don't think it's a huge deal right now if it's challenging. What's probably most valuable is actually defining a concurrency per eval.

So use case there being I have one eval that uses a model with a much lower rate limit than others ( I actually ran into this when i was doing a demo with using Anthropic ), I don't want to limit everything else because of that one eval.

Reply:
openai has different rate limits for different models too

Reply:
right that makes sense